src/components/dashboard/files/useFileActions.tsx

```diff
@@ -0,0 +1,151 @@
+import { Button, Typography } from '@mui/material';
+import { uniq } from 'lodash-es';
+import { useCallback, useMemo, useState } from 'react';
+import { Link as ReactRouterLink } from 'react-router-dom';
+
+import ConfirmationDialog from '@/components/elements/ConfirmationDialog';
+import LoadingButton from '@/components/elements/LoadingButton';
+import useSafeParams from '@/hooks/useSafeParams';
+import { useHasClientPermissions } from '@/modules/permissions/useHasPermissionsHooks';
+import { DashboardRoutes } from '@/routes/routes';
+import {
+  FileFieldsFragment,
+  useDeleteClientFileMutation,
+  useGetClientFilesQuery,
+} from '@/types/gqlTypes';
+import generateSafePath from '@/utils/generateSafePath';
+
+export type UseFileActionsArgs = {
+  onDeleteFile?: (id: string) => any;
+};
+
+const useFileActions = ({ onDeleteFile = () => {} }: UseFileActionsArgs) => {
+  const { clientId } = useSafeParams() as { clientId: string };
+  const [deletingIds, setDeletingIds] = useState<string[]>([]);
+  const [fileToDelete, setFileToDelete] = useState<
+    FileFieldsFragment | undefined
+  >();
+  const [deleting, setDeleting] = useState<boolean>(false);
+
+  const [canEdit] = useHasClientPermissions(clientId, [
+    'canManageAnyClientFiles',
+    'canManageOwnClientFiles',
+  ]);
+  const [canEditAny] = useHasClientPermissions(clientId, [
+    'canManageAnyClientFiles',
+  ]);
+  const { refetch } = useGetClientFilesQuery({ variables: { id: clientId } });
+
+  const [deleteFile] = useDeleteClientFileMutation();
+
+  const handleDeleteFile = useCallback(
+    (fileId: string) => {
+      setDeletingIds((ids) => uniq([...ids, fileId]));
+      setDeleting(true);
+      return deleteFile({
+        variables: { input: { fileId } },
+      }).finally(() => {
+        // * Using refetch here instead of eviction because it prevents a table load after delete and makes it easier to set deleting ids correctly
```

Comment:
I'm not sure what you're after here. The table _should_ reload after delete. Can we use a cache eviction here instead?

The behavior I'm seeing is: I delete an item, and it remains in the table. I need to refresh the page makes the row disappear.

Reply:
That seems like the wrong behavior, but I'm not seeing that behavior locally. Regardless, there's what I'm on about here:

These things have to happen once the file is deleted:

1. Have to reload the items in the table
2. Have to remove the deleted ID from `deletingIds`, which is what makes the loading indicator show while deleting
3. Have to set `fileToDelete` to undefined, which hides the confirm deletion modal
4. Have to set `deleting` to false, which clears the loading state on the confirmation modal
5. Have to call `onDeleteFile`, which is a callback the consumer can pass in to fire when the file is done deleting

Either an eviction or a refetch will reload the items in the table, but I think a refetch is better here. It gives me a much better place to do all the things above. With a refetch, the logic becomes:

```
deleteFileMutation -- finally --> refetch -- then --> do items 2-5 above
```

This means that when I do items 2-5, I know for sure that that file mutation has fired, returned, and the items have been refetched, which means that the operation has definitively been completed. Further, the table won't hit a spinner, it will just remove the deleted item once things are refetched.

If I use an eviction, then it will cause some number of queries to refetch, but I have no introspection into when that will happen and whether or not it's been completed. If I do items 2-5 above before things are refetched, then I'll clear all the loading state of the deletion before I've pulled new information, which could cause the indicators to behave strangely. It will also cause the table to get a full-reload with a spinner, which I think is slightly worse UX.

I could deal with this by adding a `useEffect` that checks file IDs against the previous file IDs values and does items 2-5 based on that comparison, but it feels more fragile to me. I could probably mitigate some of that with cache manipulation, but I'd rather not open that can of worms if I don't have to.

Hopefully that helps explain my logic here. I'm happy to adjust this in whatever way you'd like if you don't agree with my reasoning. Let me know if that's the case and I'll go ahead and make those changes.

Reply:
Try with pagination and see if you can reproduce the bug. You can pass `defaultPageSize={2}` to the `GenericTableWithData` in `AllFiles` to test that. If you want to use refetch that's fine with me, but it should work on a paginated table.

For all the other record tables, we use cache eviction to handle record deletion and creation. I agree it's not optimal to show the spinner. If this is a better pattern that works with pagination, we can use it everywhere. (No need to do that as part of this PR, just saying - I'd prefer if things were consistent).

Reply:
Ah, gotcha. Yeah, pagination complicates the refetch enough to make it not worth using. I've switched to eviction. Apologies, I failed to take the pagination facet into account.

Reply:
Alright, this works. It doesn't actually reload the table. In other places we would do `cache.evict({ id: Client:${clientId}, fieldName: 'files' });` instead of `cache.evict({ id: File:${id} });`, so that the whole table reloads and the pagination number is correct, but I think I like this better. The total count is off, but there's no reload. I think its a reasonable tradeoff.

![Kapture 2023-03-28 at 12 32 23](https://user-images.githubusercontent.com/5333982/228308952-fea8db48-0b44-41e9-b13f-4343b0243a50.gif)